official/mnist/mnist.py

```diff
@@ -75,7 +78,16 @@ def __call__(self, inputs, training):
     return self.fc2(y)
 
 
-def model_fn(features, labels, mode, params):
+################################################################################
+# Training loop for the model
+################################################################################
+def model_fn_with_optimizer(features, labels, mode, params):
+  """Wrapper for the model_fn that sets the optimizer for single GPU/CPU.
+  """
+  return model_fn(features, labels, mode, params, tf.train.AdamOptimizer)
```

Comment:
I'm not sure about this `model_fn_with_optimizer` refactoring. For one, this isn't convenient when trying out different optimizers or configurations as it is effectively defining an unnamed narrow "interface" for optimizer factories. For example, if I want to try adam with a different value of [`beta1` or `beta2`](https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer), then this becomes even more involved with something like:

```
return model_fn(features, labels, mode, params, lambda x: tf.train.AdamOptimizer(beta1=0.9, learning_rate=x))
```

Similar, if you want to use `MomentumOptimizer` and providing the value of momentum.

Keeping the single `model_fn` from before seems much easier to follow and change.
All that would change would be that:

```python
if mode == tf.estimator.ModeKeys.TRAIN:
  optimizer = tf.train.AdamOptimizer(learning_rate=1e-4)
  ...
```

becomes:

```python
if mode == tf.estimator.ModeKeys.TRAIN:
  optimizer = tf.train.AdamOptimizer(learning_rate=1e-4)
  if param.has_key('experimetal_multi_gpu'): # Or some other name that is appropriate
    optimizer = tf.contrib.estimator.TowerOptimizer(optimizer)
  ...
```

With the current formulation, comparing single and multi-GPU behaviors on different optimizers means you have to change the optimizer in two places and make sure they are in sync.

I think it will be valuable for this file to be as succinct as possible with as few abstractions as possible so that it is easy to identify a the few lines of code to change (e.g. optimizer configurations, data format etc.)

Reply:
I was going to wrap the grabbing of the AdamOptimizer in a shared function but I figured you would push back on that when there were only two uses :) This becomes useful, however, when looking at Resnet, where the getting of an optimizer is the same for cifar and imagenet, but the containing function is different. So rather than do it in two places there, it makes sense to extract. And then it seems wise to keep the behavior consistent across the models. This has the side effect of keeping the main processing simpler by at least one conditional, which I find useful from the perspective of a learner who wants to understand what is the core of what the main loop is actually doing (versus config details). More than happy to define a get_optimizer that returns the AdamOptimizer, which would then be the one place to pick and choose your optimizer.

Reply:
I'm confused about what this has to do with ResNet.  Is something about this going to be shared between the two?